### DISCUSSION TITLE
Copy ALL environments from one ShotGrid project to another

### DISCUSSION THREAD

**Discussion by fabiangeisler** [cpenv/cpenv#56] [Ideas]
It is nice to have the ability to copy over environments between projects in ShotGrid as it is implemented right now in the `tk-cpenv` GUI, but it can become tedious if you have a lot of environments. I think it would be nice to have some sort of template project where all "default" environments live and as soon as there is a new project you can simply copy everything over from there and be good to go. 
I know that this probably easily achievable with a small custom script, but I thought it is a nice addition to `tk-cpenv`. 🙂

**Comment by danbradham** [comment]
Yeah that's true. At the moment I use our SG event daemon. When a new project is created we copy environments from the new project's template.

I wonder what the best implementation would be for this inside tk-cpenv? We could have the before_app_launch hook check for environments in a configured template project if there is none defined for your current project. Or there could be a new engine_init hook that actually copies environments from a template project. This is tricky though because both are contingent on actually loading the config once.

**Reply by fabiangeisler** [reply]
> We could have the before_app_launch hook check for environments in a configured template project if there is none defined for your current project.

That'd be interesting in a way, that the hook would not even have to copy over the environments in order to use them. So the template project is treated as a global fall back. For a few simple environments that would be pretty helpful. For some DCCs I only have a handful of environment variables that do not change per project (like license servers or the like), so having an environment per project seems somewhat like an overkill. But if the hook would fall back to a global definition on some predefined project, that would make this much simpler. 
Also I think the actual "right" implementation for this would be to make the "Environment" entity a non-project entity. You could than have a "project" field on this entity and if it is not filled it is a "global" environment. Certainly the most painful to add though 😅 .
